File: php-node/app/Service/EmailService.php
<?php

declare(strict_types=1);

namespace App\Service;

/**
 * 邮件服务 - SMTP 发送
 */
class EmailService
{
    /**
     * 发送验证码邮件
     */
    public function sendVerificationCode(string $to, string $code, int $ttlMinutes = 5): bool
    {
        $subject = sprintf('[%s] 邮箱验证码', APP_NAME);
        $body = $this->renderCodeTemplate($code, $ttlMinutes);

        return $this->send($to, $subject, $body);
    }

    /**
     * 发送密码重置邮件
     */
    public function sendPasswordResetCode(string $to, string $code, int $ttlMinutes = 5): bool
    {
        $subject = sprintf('[%s] 密码重置验证码', APP_NAME);
        $body = $this->renderResetCodeTemplate($code, $ttlMinutes);

        return $this->send($to, $subject, $body);
    }

    /**
     * 发送邮件
     */
    public function send(string $to, string $subject, string $body, bool $isHtml = true): bool
    {
        if (empty(SMTP_HOST) || empty(SMTP_USER)) {
            // SMTP 未配置，记录日志
            $this->log("SMTP 未配置，跳过发送邮件到 {$to}: {$subject}");
            return false;
        }

        $headers = [];
        $headers[] = 'From: ' . $this->formatAddress(SMTP_FROM, SMTP_FROM_NAME);
        $headers[] = 'Reply-To: ' . $this->formatAddress(SMTP_FROM, SMTP_FROM_NAME);
        $headers[] = 'X-Mailer: PHP/' . PHP_VERSION;

        if ($isHtml) {
            $headers[] = 'MIME-Version: 1.0';
            $headers[] = 'Content-Type: text/html; charset=UTF-8';
        }

        // 使用 PHP mail() 函数或 SMTP 直连
        // 生产环境建议使用 phpmailer 等库，此处提供基础实现
        $sent = mail($to, $subject, $body, implode("\r\n", $headers));

        if (!$sent) {
            $this->log("邮件发送失败: {$to} - {$subject}");
        }

        return $sent;
    }

    /**
     * 通过 SMTP 直连发送邮件
     */
    public function sendSmtp(string $to, string $subject, string $body): bool
    {
        if (empty(SMTP_HOST)) {
            return false;
        }

        $socket = @fsockopen(
            SMTP_HOST,
            SMTP_PORT,
            $errno,
            $errstr,
            10
        );

        if (!$socket) {
            $this->log("SMTP 连接失败: {$errno} - {$errstr}");
            return false;
        }

        try {
            $this->smtpRead($socket);

            $this->smtpSend($socket, 'EHLO ' . gethostname());
            $this->smtpRead($socket);

            if (SMTP_ENCRYPTION === 'starttls') {
                $this->smtpSend($socket, 'STARTTLS');
                $this->smtpRead($socket);
                stream_socket_enable_crypto($socket, true, STREAM_CRYPTO_METHOD_TLS_CLIENT);
                $this->smtpSend($socket, 'EHLO ' . gethostname());
                $this->smtpRead($socket);
            }

            if (!empty(SMTP_USER) && !empty(SMTP_PASS)) {
                $this->smtpSend($socket, 'AUTH LOGIN');
                $this->smtpRead($socket);
                $this->smtpSend($socket, base64_encode(SMTP_USER));
                $this->smtpRead($socket);
                $this->smtpSend($socket, base64_encode(SMTP_PASS));
                $this->smtpRead($socket);
            }

            $this->smtpSend($socket, 'MAIL FROM: <' . SMTP_FROM . '>');
            $this->smtpRead($socket);
            $this->smtpSend($socket, 'RCPT TO: <' . $to . '>');
            $this->smtpRead($socket);
            $this->smtpSend($socket, 'DATA');
            $this->smtpRead($socket);

            $message = $this->buildMessage($to, $subject, $body);
            $this->smtpSend($socket, $message . "\r\n.");
            $this->smtpRead($socket);

            $this->smtpSend($socket, 'QUIT');

            return true;
        } finally {
            fclose($socket);
        }
    }

    /**
     * 构建邮件消息
     */
    private function buildMessage(string $to, string $subject, string $body): string
    {
        $boundary = md5((string) time());

        $headers = [];
        $headers[] = 'From: ' . $this->formatAddress(SMTP_FROM, SMTP_FROM_NAME);
        $headers[] = 'To: ' . $to;
        $headers[] = 'Subject: =?UTF-8?B?' . base64_encode($subject) . '?=';
        $headers[] = 'MIME-Version: 1.0';
        $headers[] = 'Content-Type: multipart/alternative; boundary="' . $boundary . '"';
        $headers[] = 'Date: ' . date('r');

        $message = implode("\r\n", $headers) . "\r\n\r\n";
        $message .= "--{$boundary}\r\n";
        $message .= "Content-Type: text/html; charset=UTF-8\r\n";
        $message .= "Content-Transfer-Encoding: base64\r\n\r\n";
        $message .= chunk_split(base64_encode($body)) . "\r\n";
        $message .= "--{$boundary}--";

        return $message;
    }

    private function smtpSend($socket, string $data): void
    {
        fwrite($socket, $data . "\r\n");
    }

    private function smtpRead($socket): string
    {
        $response = '';
        while ($line = fgets($socket, 515)) {
            $response .= $line;
            if (substr($line, 3, 1) === ' ') {
                break;
            }
        }
        return $response;
    }

    /**
     * 格式化邮件地址
     */
    private function formatAddress(string $email, string $name = ''): string
    {
        if ($name !== '') {
            return '=?UTF-8?B?' . base64_encode($name) . '?= <' . $email . '>';
        }
        return $email;
    }

    /**
     * 渲染验证码模板
     */
    private function renderCodeTemplate(string $code, int $ttlMinutes): string
    {
        return <<<HTML
<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"></head>
<body style="font-family: sans-serif; max-width: 600px; margin: 0 auto; padding: 20px;">
    <div style="background: #f5f5f5; border-radius: 8px; padding: 30px; text-align: center;">
        <h2 style="color: #333;">邮箱验证码</h2>
        <p style="color: #666; font-size: 14px;">您正在进行邮箱验证，验证码为：</p>
        <div style="font-size: 32px; font-weight: bold; color: #1976d2; letter-spacing: 8px; margin: 20px 0;">
            {$code}
        </div>
        <p style="color: #999; font-size: 12px;">验证码有效期为 {$ttlMinutes} 分钟，请勿泄露给他人。</p>
    </div>
</body>
</html>
HTML;
    }

    /**
     * 渲染密码重置验证码模板
     */
    private function renderResetCodeTemplate(string $code, int $ttlMinutes): string
    {
        return <<<HTML
<!DOCTYPE html>
<html>
<head><meta charset="UTF-8"></head>
<body style="font-family: sans-serif; max-width: 600px; margin: 0 auto; padding: 20px;">
    <div style="background: #f5f5f5; border-radius: 8px; padding: 30px; text-align: center;">
        <h2 style="color: #333;">密码重置验证码</h2>
        <p style="color: #666; font-size: 14px;">您正在重置密码，验证码为：</p>
        <div style="font-size: 32px; font-weight: bold; color: #d32f2f; letter-spacing: 8px; margin: 20px 0;">
            {$code}
        </div>
        <p style="color: #999; font-size: 12px;">验证码有效期为 {$ttlMinutes} 分钟。如非本人操作，请忽略此邮件。</p>
    </div>
</body>
</html>
HTML;
    }

    /**
     * 记录日志
     */
    private function log(string $message): void
    {
        $logDir = LOG_PATH;
        if (!is_dir($logDir)) {
            @mkdir($logDir, 0755, true);
        }

        $logFile = $logDir . '/' . date('Y-m-d') . '.log';
        @file_put_contents(
            $logFile,
            sprintf("[%s] [Email] %s\n", date('Y-m-d H:i:s'), $message),
            FILE_APPEND | LOCK_EX
        );
    }
}

File: app/services/email_service.py
"""
邮件发送服务

支持同步 SMTP 发送，可通过 run_in_executor 在异步上下文中使用。
"""
import asyncio
import logging
import smtplib
import ssl
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from typing import Optional

from app.config import get_settings

logger = logging.getLogger(__name__)


def _get_smtp_config() -> dict:
    """从系统配置获取 SMTP 设置"""
    settings = get_settings()
    return {
        "host": settings.smtp_host,
        "port": settings.smtp_port,
        "username": settings.smtp_username,
        "password": settings.smtp_password,
        "from_email": settings.smtp_from_email,
        "use_tls": settings.smtp_use_tls,
    }


def _is_smtp_configured() -> bool:
    """检查 SMTP 是否已配置"""
    config = _get_smtp_config()
    return bool(config["host"] and config["username"] and config["password"] and config["from_email"])


def _send_email_sync(
    to: str | list[str],
    subject: str,
    body: str,
    html_body: Optional[str] = None,
    from_email: Optional[str] = None,
) -> bool:
    """
    同步发送邮件

    Args:
        to: 收件人地址（单个或列表）
        subject: 邮件主题
        body: 纯文本正文
        html_body: HTML 正文（可选）
        from_email: 发件人地址（可选，默认使用系统配置）

    Returns:
        bool: 是否发送成功
    """
    config = _get_smtp_config()

    if not _is_smtp_configured():
        logger.warning("SMTP 未配置，无法发送邮件")
        return False

    if isinstance(to, str):
        to = [to]

    sender = from_email or config["from_email"]

    msg = MIMEMultipart('alternative')
    msg['From'] = sender
    msg['To'] = ', '.join(to)
    msg['Subject'] = subject

    part_text = MIMEText(body, 'plain', 'utf-8')
    msg.attach(part_text)

    if html_body:
        part_html = MIMEText(html_body, 'html', 'utf-8')
        msg.attach(part_html)

    try:
        server = smtplib.SMTP(config["host"], config["port"])
        try:
            server.ehlo()

            if config["use_tls"]:
                context = ssl.create_default_context()
                server.starttls(context=context)
                server.ehlo()

            server.login(config["username"], config["password"])
            server.sendmail(sender, to, msg.as_string())
            logger.info("邮件发送成功: %s -> %s", subject, to)
            return True
        finally:
            try:
                server.quit()
            except smtplib.SMTPServerDisconnected:
                pass
    except Exception as e:
        logger.error("邮件发送失败: %s", str(e))
        return False


async def send_email(
    to: str | list[str],
    subject: str,
    body: str,
    html_body: Optional[str] = None,
    from_email: Optional[str] = None,
) -> bool:
    """
    异步发送邮件（在线程池中执行同步 SMTP 操作）

    Args:
        to: 收件人地址
        subject: 邮件主题
        body: 纯文本正文
        html_body: HTML 正文
        from_email: 发件人地址

    Returns:
        bool: 是否发送成功
    """
    loop = asyncio.get_event_loop()
    return await loop.run_in_executor(
        None,
        _send_email_sync,
        to,
        subject,
        body,
        html_body,
        from_email,
    )


def send_email_sync(
    to: str | list[str],
    subject: str,
    body: str,
    html_body: Optional[str] = None,
    from_email: Optional[str] = None,
) -> bool:
    """
    同步发送邮件（Huey 任务使用）

    Args:
        to: 收件人地址
        subject: 邮件主题
        body: 纯文本正文
        html_body: HTML 正文
        from_email: 发件人地址

    Returns:
        bool: 是否发送成功
    """
    return _send_email_sync(to, subject, body, html_body, from_email)


async def send_verification_code(to: str, code: str) -> bool:
    """
    发送验证码邮件

    Args:
        to: 收件人地址
        code: 验证码

    Returns:
        bool: 是否发送成功
    """
    subject = "验证码 - 2c2a"
    body = f"您的验证码是: {code}\n\n验证码有效期为5分钟，请勿泄露给他人。"
    html_body = f"""
    <div style="max-width: 480px; margin: 0 auto; padding: 24px; font-family: sans-serif;">
        <h2 style="color: #333;">2c2a 验证码</h2>
        <p>您的验证码是：</p>
        <div style="font-size: 32px; font-weight: bold; letter-spacing: 4px;
                    color: #0066cc; padding: 12px 24px;
                    background: #f0f4ff; border-radius: 8px;
                    display: inline-block; margin: 8px 0;">
            {code}
        </div>
        <p style="color: #666; font-size: 14px;">
            验证码有效期为5分钟，请勿泄露给他人。
        </p>
    </div>
    """
    return await send_email(to, subject, body, html_body)


def send_verification_code_sync(to: str, code: str) -> bool:
    """
    同步发送验证码邮件（Huey 任务使用）

    Args:
        to: 收件人地址
        code: 验证码

    Returns:
        bool: 是否发送成功
    """
    subject = "验证码 - 2c2a"
    body = f"您的验证码是: {code}\n\n验证码有效期为5分钟，请勿泄露给他人。"
    return _send_email_sync(to, subject, body)
